src/Framework/Analyzer/src/WebApplicationBuilder/WebApplicationBuilderAnalyzer.cs

```diff
@@ -0,0 +1,182 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers.WebApplicationBuilder;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class WebApplicationBuilderAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(new[]
+    {
+        DiagnosticDescriptors.DoNotUseConfigureWebHostWithConfigureHostBuilder,
+        DiagnosticDescriptors.DoNotUseConfigureWithConfigureWebHostBuilder,
+        DiagnosticDescriptors.DoNotUseUseStartupWithConfigureWebHostBuilder,
+    });
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationStartAction(compilationStartAnalysisContext =>
+        {
+            var compilation = compilationStartAnalysisContext.Compilation;
+            if (!WellKnownTypes.TryCreate(compilation, out var wellKnownTypes))
+            {
+                Debug.Fail("One or more types could not be found. This usually means you are bad at spelling C# type names.");
+                return;
+            }
+
+            INamedTypeSymbol[] configureTypes = { wellKnownTypes.WebHostBuilderExtensions };
```

Comment:
Couldn't statically cache as the compilation is needed to get the well-known types, but I guess the lifetime of the class is "long enough" that these locals being reused in the lambda are OK?